DatabindingWithNewsApiKotlin/app/src/main/java/com/enpassio/databindingwithnewsapikotlin/data/NewsRepository.kt

```diff
@@ -0,0 +1,65 @@
+package com.enpassio.databindingwithnewsapikotlin.data
+
+import android.arch.lifecycle.LiveData
+import android.arch.lifecycle.MutableLiveData
+import android.content.Context
+import android.os.AsyncTask
+import android.util.Log
+import com.enpassio.databindingwithnewsapikotlin.utils.fetchArticles
+import com.enpassio.databindingwithnewsapikotlin.utils.thereIsConnection
+
+
+class NewsRepository private constructor(private val mListener: NetworkStateListener) {
+
+    /*We need a mutable live data here, so that we can modify it, but we
+    prefer to pass an immutable live data to the UI layer*/
+    private val _articles = MutableLiveData<List<Article>>()
+    val articles: LiveData<List<Article>>
+        get() = _articles
+
+    fun checkConnectionAndStartFetching(context : Context) {
+        //If data is already there, no need to go over this process again
+        if (articles.value?.isNotEmpty() == true) {
+            return
+        }
+        if (thereIsConnection(context)) {
+            //Pass network state to fragment
+            mListener.onNetworkStateChanged(true)
+            Log.d(TAG, "there is connection, start fetching")
+            //Start fetching from the News Api in a background thread
+            NewsAsyncTask().execute()
+        } else {
+            Log.d(TAG, "there is no connection")
+            //Pass network state to fragment
+            mListener.onNetworkStateChanged(false)
+        }
+    }
+
+    private inner class NewsAsyncTask : AsyncTask<Void, Void, List<Article>>() {
```

Comment:
I'm not able to re-create the issue always but while testing the app, I found this: 
```
FATAL EXCEPTION: main
    Process: com.enpassio.databindingwithnewsapikotlin, PID: 5400
    java.lang.IllegalArgumentException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkParameterIsNotNull, parameter list
        at com.enpassio.databindingwithnewsapikotlin.data.NewsRepository$NewsAsyncTask.onPostExecute(NewsRepository.kt)
        at com.enpassio.databindingwithnewsapikotlin.data.NewsRepository$NewsAsyncTask.onPostExecute(NewsRepository.kt:38)
```
If I try starting the app with no internet connection and then turn my wifi on and airplane mode off, the app crashes, **if the internet is not connected(rather just active)**. So we can also check that before making network call, we make sure that it's actually able to do network call. One of the way of doing this can be, making a ping:
```
 private var isInternetConnected = MutableLiveData<Boolean>()
    fun isInternetAvailable(): MutableLiveData<Boolean> {
        networkExecutor.execute {
            var internetConnectionStatus: Boolean
            try {
                val ipAddr = InetAddress.getByName("google.com")
                //You can replace it with your name
                internetConnectionStatus = !ipAddr.equals("")

            } catch (e: Exception) {
                internetConnectionStatus = false
            }

            if (!internetConnectionStatus) {
                val handler = Handler(Looper.getMainLooper())
                handler.post(Runnable {
                    Toast.makeText(
                        MyApplication.context,
                        "Internet not connected. Please try again!",
                        Toast.LENGTH_SHORT
                    ).show()
                })
            }
            isInternetConnected.postValue(internetConnectionStatus)
        }
        return isInternetConnected
    }
```
Ref: @link: https://stackoverflow.com/a/9570292

Reply:
@rajtheinnovator Well I can't replicate the issue but I believe you :) Which device/emulator do you use? I had come across this "pinging internet for checking connection" concept before but didn't understand much. Now it makes sense. I'll search a bit about this. 
And putting connectionStatus in a livedata is a great idea. Why did I bother with the callbacks! 
And I'll need to change this in java version as well.

Reply:
@OyaCanli I used two devices for it. OnePlus3T and a Xiaomi api 27 device. I guess the reason you're not able to replicate the issue is that by the time you make api call, internet gets connected and so it works fine. I tried running it where connectivity was not seamless and sometimes it took few seconds to get connected to internet. 

I think the code would work most of the time but yes, it can create issue if we're looking for super smooth experience, specially in production ready apps. Not critical but maybe worth trying to take a look. The code I shared above is from one of the app I was working on and there I saw its need. It worked well for my use case and I think you can try it too. Just let me know if you any issue in this or Java version :+1: